io.openems.edge.controller.asymmetric.activepowervoltagecharacteristic/src/io/openems/edge/controller/asymmetric/activepowervoltagecharacteristic/ActivePowerVoltageCharacteristicImpl.java

```diff
@@ -0,0 +1,202 @@
+		}
+
+		@Override
+		public Doc doc() {
+			return this.doc;
+		}
+	}
+
+	public ActivePowerVoltageCharacteristicImpl() {
+		super(Controller.ChannelId.values(), ChannelId.values());
+	}
+
+	@Activate
+	void activate(ComponentContext context, Config config) throws OpenemsNamedException {
+
+		super.activate(context, config.id(), config.alias(), config.enabled());
+		if (OpenemsComponent.updateReferenceFilter(cm, this.servicePid(), "ess", config.ess_id())) {
+			return;
+		}
+		if (OpenemsComponent.updateReferenceFilter(cm, this.servicePid(), "meter", config.meter_id())) {
+			return;
+		}
+		this.config = config;
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+	}
+
+	/**
+	 * Initialize the P by U characteristics.
+	 *
+	 * <p>
+	 * Gets EssID and according to that " componentId / device ", it will "point
+	 * out/ indicate/assign" on which power line connected.// Numbers will be
+	 * recursive after L3/ess3.See ess4...
+	 *
+	 * <pre>
+	 * [
+	 *  { "ess0"    : Ess Cluster}
+	 *  { "ess1"    : power line : 1}},
+	 *  { "ess2"    : power line : 2}},
+	 *  { "ess3"    : power line : 3}},
+	 *  { "ess4"    : power line : 1}},
+	 *  ......
+	 * ]
+	 * </pre>
+	 * 
+	 * @param percentQ the configured Percent-by-Q values
+	 * 
+	 * @throws OpenemsNamedException on error
+	 */
+	private essId getEssId() {
+		String[] parts = this.config.ess_id().split("ess");
+		String part2 = parts[1];
+		int essIdInt = Integer.parseInt(part2);
+		if (this.config.ess_id().equals("ess0")) {
+			return essId.ess0;
+		} else if (this.config.ess_id().equals("ess1") || ((essIdInt % 3) == 1)) {
+			return essId.ess1;
+		} else if (this.config.ess_id().equals("ess2") || ((essIdInt % 3) == 2)) {
+			return essId.ess2;
+		}
+		return essId.ess3;
+	}
+
+	private enum essId {
+		ess0, ess1, ess2, ess3
+	}
+
+	@Override
+	public void run() throws OpenemsNamedException {
+		GridMode gridMode = this.ess.getGridMode();
+		if (gridMode.isUndefined()) {
+			this.logWarn(this.log, "Grid-Mode is [UNDEFINED]");
+		}
+		switch (gridMode) {
+		case ON_GRID:
+		case UNDEFINED:
+			break;
+		case OFF_GRID:
+			return;
+		}
+
+		Channel<Integer> gridLineVoltage;
+		switch (this.getEssId()) {
+		case ess0:
+			this.log.info("\n'ess0' assumed as a essCluster by this controller.\n"
+					+ "Each ess number related with power line, ex: ess1 == L1 ");
+			return;
+		case ess1:
+			gridLineVoltage = this.meter.channel(AsymmetricMeter.ChannelId.VOLTAGE_L1);
+			this.voltageRatio = gridLineVoltage.value().orElse(0) / (this.config.nominalVoltage() * 1000);
+			break;
+		case ess2:
+			gridLineVoltage = this.meter.channel(AsymmetricMeter.ChannelId.VOLTAGE_L2);
+			this.voltageRatio = gridLineVoltage.value().orElse(0) / (this.config.nominalVoltage() * 1000);
+			break;
+		case ess3:
+			gridLineVoltage = this.meter.channel(AsymmetricMeter.ChannelId.VOLTAGE_L3);
+			this.voltageRatio = gridLineVoltage.value().orElse(0) / (this.config.nominalVoltage() * 1000);
+			break;
+		}
```

Comment:
No... just no! This seems to be very project specific. No chance you get that into `develop`! ;-)

Reply:
As you said, yes its very project specific. Sorry for that.